_posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md

```diff
@@ -0,0 +1,201 @@
+
+##### Ironic & Ironic Python Agent (IPA)
+- Ironic is a bare-metal provisioning tool that handles tasks like booting servers, installing operating systems, and configuring hardware. Think of Ironic as the piece of software that manages actual physical servers. In a Metal3 workflow, Ironic receives orders from BMO, and translates into actionable steps.
+- Ironic Python Agent (IPA) is the software that runs on each bare-metal node to carry out Ironic’s instructions. It interacts with the hardware directly, like wiping disks, configuring networks, and handling boot processes.
+
+__Quick Intro: How Ironic and IPA Work Together in Metal3__
+Ironic is the bare-metal provisioning tool used in Metal3. Think of it as the system that manages actual servers—turning them on, installing operating systems, and so on. When you want to deploy a Kubernetes cluster on bare-metal hardware, Ironic is the one making sure that each physical server is up and running with the right setup.
+
+The Ironic Python Agent (IPA) plays a super important role here. It’s a little program that runs on the bare-metal machines Ironic is provisioning. You can think of IPA as the "hands" of Ironic—it actually touches the hardware to do things like:
+
+- Booting up the machine.
+- Installing the operating system.
+- Configuring network settings.
+
+In a normal Metal3 workflow, BMO translates the kubernetes reconciling logic to concrete actions and forward them to Ironic, which tells the IPA what to do. For example, BMO might notice that  Ironic might tell IPA to boot up a server and install an image, and IPA will carry out these tasks. Once the node is ready, Ironic passes control back to CAPI and CAPM3, which can then set up Kubernetes clusters on these provisioned machines.
+
+
+## The Experiment: Simulating 1000 Kubernetes Clusters
+The goal of this experiment was to push Metal3 to simulate 1000 single-node Kubernetes clusters on fake hardware. Instead of provisioning real machines, we used Fake Ironic Python Agents (Fake IP) and Fake Kubernetes API Servers (FKAS) to simulate nodes and control planes, respectively. This setup allowed us to test a massive environment without the need for physical infrastructure.
+
+### Step 1: Setting Up the environment
+As you may have known, a typical Metal3 workflow requires existence of several components: bootstrap kubernetes cluster, special networks, baremetal nodes, etc. As we are working on simulating environment, we will start with a newly spawn Ubuntu VM, create a cluster with minikube, add networks with libvirt, and so on (If you're familiar with the Metal3's dev-env, this step is similar to what script [01](), [02] and a part of [03] do). We will not go into details on this part, but you can find the related setup from [this script]() if you are interested.
```

Comment:
"special networks", not really, we use separate provisioning network but it is not a hard requirement.

Reply:
I think ironic requires `baremetal` network, doesn't it?

Reply:
e.g in a scenario where we use refish with virtual media booting and there is no network seapration between the NICs including the BMC NIC we can use 1 network.